app/(candidate)/onboarding/[slug]/[step]/components/ballotOffices/BallotRaces.js

```diff
@@ -59,57 +79,63 @@ export default function BallotRaces(props) {
     }
   };
 
+  const clearState = () => {
+    setSelected(false);
+    setInputValue('');
+    setLevel('')
+    setYearFilter(null)
+  }
+
   const handleZipChange = async (newZip) => {
     if (newZip !== zip) {
-      setLoading(true);
-      const newRaces = await fetchRaces(newZip);
-      setRaces(newRaces.races);
-      setFilteredRaces(newRaces.races);
-      setZip(newZip);
-      setSelected(false);
-      setLoading(false);
-      setFilters('');
-      const updatedCampaign = {
+      setZip(newZip)
+      clearState()
+      await loadRaces(newZip)
+      await updateCampaign({
         ...campaign,
         details: {
           ...campaign.details,
           zip: newZip,
         },
-      };
-      await updateCampaign(updatedCampaign);
+      });
     }
   };
 
-  const filterRaces = (value, updatedLevel) => {
-    if (!value && updatedLevel === '') {
-      setFilteredRaces(races);
-      return;
+  const filterRace = (race) => {
+    let positionLevel = race.position.level;
+
+    // TODO: move this kind of data-rewite closer to "source of truth", preferably
+    //  whatever entry point the data is being ingested into our database
+    if (positionLevel === 'CITY') {
+      positionLevel = 'LOCAL';
     }
-    if (races && typeof races.filter === 'function') {
-      const filtered = races.filter((option) => {
-        let positionLevel = option.position.level;
-        if (positionLevel === 'CITY') {
-          positionLevel = 'LOCAL';
-        }
+    const raceYear = (new Date(race.election.electionDay)).getFullYear()
 
-        if (value && (!updatedLevel || updatedLevel === '')) {
-          return option.position.name
-            .toLowerCase()
-            .includes(value.toLowerCase());
-        }
-        if (value && updatedLevel && updatedLevel !== '') {
-          return (
-            option.position.name.toLowerCase().includes(value.toLowerCase()) &&
-            positionLevel === updatedLevel.toUpperCase()
-          );
-        }
+    const textFiltered = inputValue ?
+      race.position.name
+      .toLowerCase()
+      .includes(inputValue.toLowerCase()) : true
 
-        return positionLevel === updatedLevel.toUpperCase();
-      });
-      setFilteredRaces(filtered);
-    }
-  };
+    const levelFiltered = level ?
+      level && positionLevel === level.toUpperCase() :true
+
+    const yearFiltered = yearFilter ?
+      parseInt(yearFilter) === raceYear : true
```

Comment:
not a big deal, I like "is" in boolean variables. `isLeveliltered` `isYearFiltered`